File: 2017/day13/day13.py
from typing import Optional, List, Tuple


class Scanner:
    def getInput(self) -> List[Tuple[int]]:
        inputFile = "input-test.txt" if self.useTest else "input.txt"
        lasers = []
        with open(inputFile, "r") as file1:
            for line in file1:
                line = line.strip().split(": ")
                lasers.append((int(line[0]), int(line[1])))
        return lasers

    def __init__(self, useTest: Optional[bool] = False) -> None:
        self.useTest = useTest
        self.lasers = self.getInput()

    def getSeverity(self, delay: Optional[int] = 0) -> int:
        severity = 0
        for index, length in self.lasers[0:]:
            if (index + delay) % (length * 2 - 2) == 0:
                if delay != 0: # for Part 2
                    return 1
                severity += length * index
        return severity

    def getMinDelay(self) -> int:
        delay = 0
        while delay < 10_000_000 and self.getSeverity(delay) > 0:
            delay += 1
        return delay


if __name__ == "__main__":
    scanner = Scanner()
    print("Day 13 part 1:", scanner.getSeverity())
    print("Day 13 part 2:", scanner.getMinDelay())

File: 2016/day13/day13.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

type Office struct {
	UseTest      bool
	OfficeNum    int
	Dest         []int
	MaxLocations int
	Steps        int
}

func (this *Office) getInput() {
	inputFile := "input.txt"
	if this.UseTest {
		inputFile = "input-test.txt"
		this.Dest = []int{7, 4}
	} else {
		this.Dest = []int{31, 39}
	}
	file, _ := os.Open(inputFile)
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		this.OfficeNum, _ = strconv.Atoi(line)
	}
	defer file.Close()
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func (this *Office) findLocation(startPos []int) {
	// BFS
	dirMap := [][]int{{0, -1}, {1, 0}, {0, 1}, {-1, 0}}

	visited := make(map[string]bool)
	var queue [][]int
	queue = append(queue, startPos)

	steps := 0

	for len(queue) > 0 {
		if steps == 50 {
			this.MaxLocations = len(visited)
		}
		temp := [][]int{}

		for _, currPos := range queue {
			x, y := currPos[0], currPos[1]
			if x == this.Dest[0] && y == this.Dest[1] {
				this.Steps = steps
				return
			}

			for _, dir := range dirMap {
				newX, newY := dir[0]+x, dir[1]+y
				isAWall := isWall(newX, newY, this.OfficeNum)
				key := strconv.Itoa(newX) + "," + strconv.Itoa(newY)
				_, exists := visited[key]

				if exists {
					continue
				}
				if !isAWall {
					visited[key] = true
					temp = append(temp, []int{newX, newY})
				}
			}
		}
		queue = temp
		steps++
	}
	return
}

func isWall(x int, y int, officeNum int) bool {
	if x < 0 || y < 0 {
		return true
	}
	val := x*x + 3*x + 2*x*y + y + y*y + officeNum
	count := 0
	for val != 0 {
		count += val & 1
		val >>= 1
	}
	return count%2 == 1
}

func (this *Office) getSteps() int {
	return this.Steps
}

func (this *Office) getMaxLocations() int {
	return this.MaxLocations
}

func main() {
	office := &Office{
		UseTest: false,
	}
	office.getInput()
	office.findLocation([]int{1, 1})
	fmt.Println("Day 13 part 1:", office.getSteps())
	fmt.Println("Day 13 part 2:", office.getMaxLocations())
}
